File: app/src/main/java/com/major/qr/dialog/UploadDialog.java
package com.major.qr.dialog;

import static android.app.Activity.RESULT_OK;
import static com.major.qr.ui.LoginActivity.URL;

import android.Manifest;
import android.app.AlertDialog;
import android.app.Dialog;
import android.content.ActivityNotFoundException;
import android.content.Intent;
import android.database.Cursor;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.os.Environment;
import android.provider.MediaStore;
import android.provider.Settings;
import android.util.Log;
import android.view.View;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.core.app.ActivityCompat;
import androidx.fragment.app.DialogFragment;

import com.major.qr.R;
import com.major.qr.databinding.DialogUploadBinding;
import com.major.qr.interfaces.UpdateService;
import com.major.qr.ui.LoginActivity;
import com.major.qr.utils.FileUtils;
import com.major.qr.viewmodels.DocumentViewModel;

import java.io.File;
import java.util.Objects;

import okhttp3.MediaType;
import okhttp3.MultipartBody;
import okhttp3.RequestBody;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

public class UploadDialog extends DialogFragment {
    public static final String TAG = UploadDialog.class.getSimpleName();
    private final int FILE_PICK_REQUEST = 200;
    private final int REQUEST_CAPTURE_IMAGE = 201;
    private final DocumentViewModel documentViewModel;
    Uri uri;
    String imageFilePath;
    private DialogUploadBinding binding;
    private String documentReference;

    public UploadDialog(String documentReference, DocumentViewModel documentViewModel) {
        this.documentViewModel = documentViewModel;
        this.documentReference = documentReference;
    }

    public UploadDialog(DocumentViewModel documentViewModel) {
        this.documentViewModel = documentViewModel;
    }

    @NonNull
    @Override
    public Dialog onCreateDialog(@Nullable Bundle savedInstanceState) {
        AlertDialog.Builder builder = new AlertDialog.Builder(requireActivity());
        binding = DialogUploadBinding.inflate(getLayoutInflater());
        builder.setView(binding.getRoot()).setTitle("Select File");
        if (documentReference != null) binding.docReference.setText(documentReference);
        else binding.docReference.setVisibility(View.GONE);

        binding.fileSelect.setOnClickListener(view -> {
            String[] supportedMimeTypes = {"application/pdf", "application/msword", "image/*"};
            Intent intent = new Intent(Intent.ACTION_OPEN_DOCUMENT);
            intent.setType("*/*");
            intent.putExtra(Intent.EXTRA_MIME_TYPES, supportedMimeTypes);

            try {
                startActivityForResult(Intent.createChooser(intent, "Select a File to Upload"), FILE_PICK_REQUEST);
            } catch (ActivityNotFoundException ex) {
                // Potentially direct the user to the Market with a Dialog
                Toast.makeText(getContext(), "Please install a File Manager.", Toast.LENGTH_SHORT).show();
            }
        });

        binding.upload.setOnClickListener(view -> {
            ActivityCompat.requestPermissions(requireActivity(),
                    new String[]{
                            Manifest.permission.READ_EXTERNAL_STORAGE,
                            Manifest.permission.MANAGE_EXTERNAL_STORAGE
                    }, 1
            );
            // If you have access to the external storage, do whatever you need
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
                if (Environment.isExternalStorageManager()) {
                    // If you don't have access, launch a new activity to show the user the system's dialog
                    // to allow access to the external storage
                } else {
                    Intent intent = new Intent();
                    intent.setAction(Settings.ACTION_MANAGE_APP_ALL_FILES_ACCESS_PERMISSION);
                    Uri uri = Uri.fromParts("package", requireActivity().getPackageName(), null);
                    intent.setData(uri);
                    startActivity(intent);
                }
            }

            String path = binding.fileSelect.getText().toString();
            if (!path.isEmpty() || !path.equals(getResources().getString(R.string.file_name))) {
                String[] strings = Objects.requireNonNull(FileUtils.getPath(getContext(), uri)).split("/");
                StringBuilder stringBuilder = new StringBuilder();
                for (int i = 0; i < strings.length - 1; ++i)
                    stringBuilder.append("/").append(strings[i]);
                File file = new File(stringBuilder.toString(), strings[strings.length - 1]);
                Log.d(TAG, "onCreate: " + file.getAbsolutePath());
                if (documentReference != null) {
                    Log.d(TAG, "documentReference = " + documentReference);

                    Retrofit retrofit = new Retrofit.Builder()
                            .baseUrl(URL + "/")
                            .addConverterFactory(GsonConverterFactory.create())
                            .build();
                    UpdateService updateService = retrofit.create(UpdateService.class);
//                    MultipartBody.Part filePart = MultipartBody.Part.createFormData("document",
//                            file.getName(), RequestBody.create(MediaType.parse(
//                                    requireActivity().getContentResolver().getType(Uri.fromFile(file))), file));
                    MultipartBody.Part filePart = MultipartBody.Part.createFormData("document",
                            file.getName(), RequestBody.create(MediaType.parse("image/*"), file));
                    MultipartBody body = new MultipartBody.Builder()
                            .addFormDataPart("documentReference", documentReference).build();
                    RequestBody body1 = RequestBody.create(MediaType.parse("multipart/form-data"), documentReference);
                    Call<String> jsonObjectCall = updateService.updateDoc(LoginActivity.ACCESS_TOKEN, filePart, body1);
                    jsonObjectCall.enqueue(new Callback<String>() {
                        @Override
                        public void onResponse(Call<String> call, Response<String> response) {
                            Log.d(TAG, "call = " + response);
                        }

                        @Override
                        public void onFailure(Call<String> call, Throwable t) {
                            Log.e(TAG, "onFailure: " + call);
                        }
                    });

//                    documentViewModel.updateDoc(documentReference, file).observe(this, s -> {
//                        Toast.makeText(getContext(), "Successfully updated!", Toast.LENGTH_SHORT).show();
//                        dismiss();
//                    });
                } else documentViewModel.uploadDoc(file).observe(this, s -> {
                    Toast.makeText(getContext(), "Successfully uploaded!", Toast.LENGTH_SHORT).show();
                    dismiss();
                });
            } else
                Toast.makeText(getContext(), "Select a file first!", Toast.LENGTH_SHORT).show();
        });

        return builder.create();
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        if (requestCode == FILE_PICK_REQUEST) {
            if (resultCode == RESULT_OK) {
                // Get the Uri of the selected file
                uri = data.getData();
                Log.d(TAG, "File Uri: " + uri.toString());
                String path;
                String[] projection = {MediaStore.Files.FileColumns.DATA};
                Cursor cursor = requireActivity().getContentResolver().query(uri, projection, null, null, null);
                if (cursor == null) {
                    path = uri.getPath();
                } else {
                    cursor.moveToFirst();
                    int column_index = cursor.getColumnIndexOrThrow(projection[0]);
                    path = cursor.getString(column_index);
                    cursor.close();
                }
                if (path == null || path.isEmpty()) {
                    Log.d("uri.getPath() = ", uri.getPath());
                    binding.fileSelect.setText(uri.getPath());
                } else {
                    Log.d("path = ", path);
                    binding.fileSelect.setText(path);
                }
            }
        }
        if (requestCode == REQUEST_CAPTURE_IMAGE) {
            Log.d(requestCode + "", resultCode + "");
            if (resultCode == RESULT_OK) {
                Toast.makeText(getContext(), "Captured! Click on Upload to upload the image.",
                        Toast.LENGTH_SHORT).show();
                binding.fileSelect.setText(imageFilePath);
                uri = Uri.fromFile(new File(imageFilePath));
                Log.d(TAG, "onActivityResult: " + uri);
            }
        } else {
            super.onActivityResult(requestCode, resultCode, data);
        }
    }
}

File: app/src/main/java/com/major/qr/dialog/UploadDialog.kt
package com.major.qr.dialog

import android.Manifest
import android.app.Activity
import android.app.AlertDialog
import android.app.Dialog
import android.content.ActivityNotFoundException
import android.content.Intent
import android.graphics.Color
import android.graphics.drawable.ColorDrawable
import android.net.Uri
import android.os.Build
import android.os.Bundle
import android.os.Environment
import android.provider.MediaStore
import android.provider.Settings
import android.util.Log
import android.view.View
import android.widget.Toast
import androidx.core.app.ActivityCompat
import androidx.fragment.app.DialogFragment
import com.major.qr.R
import com.major.qr.databinding.DialogUploadBinding
import com.major.qr.dialog.UploadDialog
import com.major.qr.services.UpdateService
import com.major.qr.ui.DocumentFragment
import com.major.qr.ui.LoginActivity
import com.major.qr.utils.FileUtils
import com.major.qr.viewmodels.DocumentViewModel
import okhttp3.MediaType
import okhttp3.MultipartBody
import okhttp3.RequestBody
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory
import java.io.File

class UploadDialog : DialogFragment {
    val TAG = UploadDialog::class.java.simpleName
    private val FILE_PICK_REQUEST = 200
    private val REQUEST_CAPTURE_IMAGE = 201
    private val documentViewModel: DocumentViewModel
    var uri: Uri? = null
    var imageFilePath: String? = null
    private lateinit var binding: DialogUploadBinding
    private var documentReference: String? = null

    constructor(documentReference: String?, documentViewModel: DocumentViewModel) {
        this.documentViewModel = documentViewModel
        this.documentReference = documentReference
    }

    constructor(documentViewModel: DocumentViewModel) {
        this.documentViewModel = documentViewModel
    }

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        val builder = AlertDialog.Builder(requireActivity())
        binding = DialogUploadBinding.inflate(layoutInflater)
        builder.setView(binding.root)

        if (documentReference != null)
            binding.docReference.text = documentReference
        else
            binding.docReference.visibility = View.GONE

        binding.fileSelect.setOnClickListener { view: View? ->
            val supportedMimeTypes = arrayOf("application/pdf", "application/msword", "image/*")
            val intent = Intent(Intent.ACTION_OPEN_DOCUMENT)
            intent.type = "*/*"
            intent.putExtra(Intent.EXTRA_MIME_TYPES, supportedMimeTypes)
            try {
                startActivityForResult(
                    Intent.createChooser(intent, "Select a File to Upload"),
                    FILE_PICK_REQUEST
                )
            } catch (ex: ActivityNotFoundException) {
                // Potentially direct the user to the Market with a Dialog
                Toast.makeText(context, "Please install a File Manager.", Toast.LENGTH_SHORT).show()
            }
        }
        binding.upload.setOnClickListener { view: View? ->
            ActivityCompat.requestPermissions(
                requireActivity(), arrayOf(
                    Manifest.permission.READ_EXTERNAL_STORAGE,
                    Manifest.permission.MANAGE_EXTERNAL_STORAGE
                ), 1
            )
            // If you have access to the external storage, do whatever you need
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
                if (Environment.isExternalStorageManager()) {
                    // If you don't have access, launch a new activity to show the user the system's dialog
                    // to allow access to the external storage
                } else {
                    val intent = Intent()
                    intent.action = Settings.ACTION_MANAGE_APP_ALL_FILES_ACCESS_PERMISSION
                    val uri = Uri.fromParts("package", requireActivity().packageName, null)
                    intent.data = uri
                    startActivity(intent)
                }
            }

            val path = binding.fileSelect.text.toString()
            if (path.isNotEmpty() || path != resources.getString(R.string.file_name)) {
                val strings = FileUtils.getPath(requireContext(), uri!!)?.split("/".toRegex())
                    ?.dropLastWhile { it.isEmpty() }?.toTypedArray()
                val stringBuilder = StringBuilder()
                if (strings != null) {
                    for (i in 0 until strings.size - 1) stringBuilder.append("/").append(strings[i])
                }
                val file = File(stringBuilder.toString(), strings?.get(strings.size - 1) ?: "")
                Log.d(TAG, "onCreate: " + file.absolutePath)
                if (documentReference != null) {
                    Log.d(TAG, "documentReference = $documentReference")
                    val retrofit = Retrofit.Builder()
                        .baseUrl(LoginActivity.URL + "/")
                        .addConverterFactory(GsonConverterFactory.create())
                        .build()
                    val updateService = retrofit.create(UpdateService::class.java)
                    //                    MultipartBody.Part filePart = MultipartBody.Part.createFormData("document",
//                            file.getName(), RequestBody.create(MediaType.parse(
//                                    requireActivity().getContentResolver().getType(Uri.fromFile(file))), file));
                    val filePart = MultipartBody.Part.createFormData(
                        "document",
                        file.name, RequestBody.create(MediaType.parse("image/png"), file)
                    )
                    val body = documentReference?.let {
                        MultipartBody.Builder()
                            .addFormDataPart("documentReference", it).build()
                    }
                    val body1 = RequestBody.create(
                        MediaType.parse("multipart/form-data"),
                        documentReference
                    )
                    val jsonObjectCall =
                        updateService.updateDoc(LoginActivity.ACCESS_TOKEN, filePart, body1)
                    jsonObjectCall.enqueue(object : Callback<String?> {
                        override fun onResponse(call: Call<String?>, response: Response<String?>) {
                            Log.d(TAG, "call = $response")
                        }

                        override fun onFailure(call: Call<String?>, t: Throwable) {
                            Log.e(TAG, "onFailure: $call")
                        }
                    })

//                    documentViewModel.updateDoc(documentReference, file).observe(this, s -> {
//                        Toast.makeText(getContext(), "Successfully updated!", Toast.LENGTH_SHORT).show();
//                        dismiss();
//                    });
                } else documentViewModel.uploadDoc(file).observe(this) { s: String? ->
                    Toast.makeText(context, "Successfully uploaded!", Toast.LENGTH_SHORT).show()
                    val tr = requireActivity().supportFragmentManager.beginTransaction()
                    tr.replace(R.id.fragment, DocumentFragment())
                    tr.commit()
                    dismiss()
                }
            } else Toast.makeText(context, "Select a file first!", Toast.LENGTH_SHORT).show()
        }
        val dialog: Dialog = builder.create()
        dialog.window!!.setBackgroundDrawable(ColorDrawable(Color.TRANSPARENT))
        return dialog
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        if (requestCode == FILE_PICK_REQUEST) {
            if (resultCode == Activity.RESULT_OK) {
                // Get the Uri of the selected file
                uri = data!!.data
                assert(uri != null)
                Log.d(TAG, "File Uri: $uri")
                val path: String?
                val projection = arrayOf(MediaStore.Files.FileColumns.DATA)
                val cursor = requireActivity().contentResolver.query(
                    uri!!, projection, null, null, null
                )
                if (cursor == null) {
                    path = uri!!.path
                } else {
                    cursor.moveToFirst()
                    val column_index = cursor.getColumnIndexOrThrow(projection[0])
                    path = cursor.getString(column_index)
                    cursor.close()
                }
                if (path.isNullOrEmpty()) {
                    Log.d("uri.getPath() = ", uri!!.path!!)
                    binding.fileSelect.text = uri!!.path
                } else {
                    Log.d("path = ", path)
                    binding.fileSelect.text = path
                }
            }
        }
        if (requestCode == REQUEST_CAPTURE_IMAGE) {
            Log.d(requestCode.toString() + "", resultCode.toString() + "")
            if (resultCode == Activity.RESULT_OK) {
                Toast.makeText(
                    context, "Captured! Click on Upload to upload the image.",
                    Toast.LENGTH_SHORT
                ).show()
                binding.fileSelect.text = imageFilePath
                uri = Uri.fromFile(File(imageFilePath))
                Log.d(TAG, "onActivityResult: $uri")
            }
        } else {
            super.onActivityResult(requestCode, resultCode, data)
        }
    }
}